File: Practicas-propias/Agresivos.cpp
#include <bits/stdc++.h>
using namespace std;
#define MAX 100002
#define MAXN 1000000000
#define ll long long
ll v[MAX];
ll n,p;
bool solve(ll x){
	ll uso=0,uso2=0;
	for(int i=1;i<=n;i++){
		if(v[i]>=uso2){
			uso++;
			uso2=v[i]+x;
		}
	}
	if(uso>=p)return true;
	return false;
}
ll bs(ll in,ll fin){
	if(in==fin)return in;
	ll mid=(in+fin+1)/2;
	if(!solve(mid))return bs(in,mid-1);
	else return bs(mid,fin);
}
	int main (){
		ios_base::sync_with_stdio(0);
		cin.tie(0);
		cin>>n>>p;
		for(int i=1;i<=n;i++){
			cin>>v[i];
		}	
		sort(v,v+n+1);
		cout << bs(1,MAXN)<<"\n";
	}

File: Solutions/Agresivos.hs
import Data.List
import Data.Char
import Control.Applicative

slt :: [Int]->Int->Int->Int->Int
slt xs i uso uso2 = if(xs !! i)>=uso2 then uso+1 else uso

slt2 :: [Int]->Int->Int->Int->Int
slt2 xs i uso2 v = if(xs !! i)>=uso2 then xs !! i + v else uso2

solve :: Int->Int->Int->Int->Int->Int->[Int]->Int
solve uso uso2 i (-1) p v xs = if uso>=p then 1 else 0
solve uso uso2 i n p v xs = solve (slt xs i uso uso2) (slt2 xs i uso2 v) (i+1) (n-1) p v xs  

bs :: [Int]->Int->Int->Int->Int
bs xs ini fin p 
	| ini == fin = ini
	| solve 0 0 0 (length xs - 1) p mid xs == 0  = bs xs ini (mid-1) p
	| otherwise = bs xs (mid) fin p
	where mid = div (ini+fin+1) 2

agv :: Int->[Int]->Int
agv p xs = bs xs 1 (10^9) p

main = do
	[x,y]<-map read . words <$> getLine
	xs<-map read . words <$> getContents
	print (agv y (sort xs))
